### PR TITLE
x11: On Xwayland, don't disable the crtc before setting screen size

### PR DESCRIPTION
**Pull request by smcv**
If we disable the crtc we can't successfully re-enable it,
leaving Xwayland stuck in a broken state until it is restarted (which
often requires logging out and back in).

Resolves: https://github.com/libsdl-org/SDL/issues/4630

---

Alternatives to this include #4631 and #4638.

This is a lighter-weight version of #4631 which seems to work as desired on:

* Wayland + Xwayland (on Mesa, AMD GPU)
* purely X11 (on Mesa, AMD GPU)
* purely X11 (on NVIDIA proprietary driver)

@amshafer, does this seem reasonable to you, or is there a reason I'm missing why disabling the crtc is important?

### CODE REVIEW

**Comment by amshafer** [comment]
tl;dr yes I think it is actually necessary, at least for best adherence to the spec and fixing Metro.

The RandR spec [1] specifies that the CRTC must fit inside the screen size. With metro as an example, it might grow the screen size smaller than what the CRTC reports, which would cause a badmatch error. The spec mentions disabling the CRTC to avoid this.

```
This request sets the CRTC to the specified position, mode, rotation
and reflection. The entire area of the CRTC must fit within the
screen size, else a Match error results. As an example, rotating the
screen so that a single CRTC fills the entire screen before and
after may necessitate disabling the CRTC, resizing the screen,
then re-enabling the CRTC at the new configuration to avoid an
invalid intermediate configuration.
```

I retested your change on Metro Exodus and still saw hangs while changing the resolution.

I'm not well-read on the internals of Xwayland and I don't really want to toss blame around without certainty, but this sounds like an issue with their handling of screens/crtcs? My change should adhere to the spec (assuming I've read the spec correctly, corrections welcome), so I would expect it to work under Xwayland. Adding a workaround if xwayland is detected could work in the meantime, but seems like checking in on why xwayland doesn't handle it correctly would be the ideal solution?

[1] https://cgit.freedesktop.org/xorg/proto/randrproto/tree/randrproto.txt  (towards the end of the RRSetCrtcConfig section)

**Comment by smcv** [comment]
> I'm not well-read on the internals of Xwayland and I don't really want to toss blame around without certainty, but this sounds like an issue with their handling of screens/crtcs?

Quite possibly: I've been able to reproduce a similar thing with just `xrandr(1)`, which suggests that something is not happening correctly in Xwayland.

However, Xwayland is part of the operating system and not trivial to replace (in LTS distributions it's essentially frozen for 2 years), whereas we want to be able to include an updated SDL in things like games and the Steam Runtime; so I think this at least needs a workaround to stop existing installations from regressing.

**Comment by amshafer** [comment]
Interesting, in that case I agree a workaround is a good idea

**Comment by smcv** [comment]
#4638 is another possible route for this.

**Comment by smcv** [comment]
> I'm not well-read on the internals of Xwayland and I don't really want to toss blame around without certainty, but this sounds like an issue with their handling of screens/crtcs? My change should adhere to the spec (assuming I've read the spec correctly, corrections welcome), so I would expect it to work under Xwayland.

I agree this doesn't seem right, and I've opened https://gitlab.freedesktop.org/xorg/xserver/-/issues/1209 with a minimal reproducer. It turns out `XRRSetCrtcConfig` isn't necessary to reproduce this: disabling and immediately re-enabling the crtc is enough.

**Comment by smcv** [comment]
> If disabling the crtc is necessary for real hardware-backed X servers, another possible refinement of this would be to check whether the Xrandr output name starts with XWAYLAND (this seems to be canonical, it's the approach used in https://gitlab.freedesktop.org/xorg/app/xisxwayland) and if so, avoid disabling the crtc.

I've updated this PR to do that, but I think #4638 might be a better workaround.

**Comment by smcv** [comment]
@slouken merged #4638 which avoids the Xwayland issue, so this is no longer necessary.
